afterburner/src/main/java/com/fasterxml/jackson/module/afterburner/util/MyClassLoader.java

```diff
@@ -56,52 +64,157 @@ public static boolean canAddClassInPackageOf(Class<?> cls)
+        //      b) the only non-private method which calls this method and thus acquires this lock is
+        //          MyClassLoader#loadAndResolve,
+        //      c) nothing in the synchronized region can have the effect of calling #loadAndResolve on this
+        //          or any other instance of MyClassLoader.
+        synchronized (getParentClassLoadingLock(parentClassLoader, className.getDottedName())) {
+            // First: check to see if the parent classloader has loaded it already
+            Class<?> impl = findLoadedClassOnParent(parentClassLoader, className.getDottedName());
+            if (impl != null) {
+                return impl;
             }
-            throw new IllegalArgumentException("Failed to load class '"+className+"': "+t.getMessage(), t);
+
+            // Important: bytecode is generated with a template name (since bytecode itself
+            // is used for checksum calculation) -- must be replaced now, however
+            replaceName(byteCode, className.getSlashedTemplate(), className.getSlashedName());
+
+            // Second: define a new class instance on the parent classloder using the bytecode
+            impl = defineClassOnParent(parentClassLoader, className.getDottedName(), byteCode, 0, byteCode.length);
+            // important: must also resolve the newly-created class.
+            resolveClassOnParent(parentClassLoader, impl);
+            return impl;
+        }
+    }
+
+    /**
+     * Get the class loading lock for the parent class loader for loading the named class.
+     *
+     * This is effectively the same implementation as ClassLoader#getClassLoadingLock, but using
+     * our static parentParallelLockMap and keying off of the parent ClassLoader instance as well as
+     * the class name to load.
+     *
+     * @param parentClassLoader     The parent ClassLoader
+     * @param className             The name of the to-be-loaded class
+     */
+    private Object getParentClassLoadingLock(ClassLoader parentClassLoader, String className) {
+        // N.B. using the canonical name and identity hash code to represent the parent class loader in the key
+        // in case that ClassLoader instance (which could be anything) implements #hashCode or #toString poorly.
+        // In the event of a collision here (same key, different parent class loader), we will end up using the
+        // same lock only to synchronize loads of the same class on two different class loaders,
+        // which shouldn't ever deadlock (see proof in #loadAndResolveUsingParentClassloader);
+        // worst case is unnecessary contention for the lock.
+        String key = String.format("%s:%d:%s",
+                parentClassLoader.getClass().getCanonicalName(),
+                System.identityHashCode(parentClassLoader),
+                className);
+        Object newLock = new Object();
+        Object lock = parentParallelLockMap.putIfAbsent(key, newLock);
```

Comment:
I don't think it will work to simply expire entries from the map when the lock is released. Consider this sequence of operations contending for a lock for the same class name/loader pair (i.e. with the same map key: 'key'), beginning with an empty map:
	[Thread A]: requests lock object for 'key'; map contains no entry for 'key' so Object-1 is created and added to map, then returned.
	[Thread A]: attempts to acquire lock for Object-1; succeeds immediately.
	[Thread B]: requests lock object for 'key'; receives Object-1 from map.
	[Thread B]: attempts to acquire lock for Object-1; blocks.
	[Thread A]: completes its work in the synchronized region and then (in either order) releases the lock on Object-1 and removes the `'key' -> Object-1` mapping from the map.
	[Thread B]: immediately acquires lock for Object-1 and unblocks.
	[Thread C]: requests lock object for 'key'; map contains no entry for 'key' so Object-2 is created and added to map, then returned.
	[Thread C]: attempts to acquire lock for Object-2; succeeds immediately.
	
Now, Thread B and Thread C are both in the synchronized region together for what should be the same lock.

A Guava (or Caffeine) `Cache` with weak values would do the trick, but I doubt it's worth adding that dependency. For what it's worth, though, this is the same implementation as used by `ClassLoader` itself. Its lock map also grows without bound (but, since it only ever gains one entry per loaded class it can't really grow **that** big)

Reply:
Good call @jbagdis , having a corresponding remove does allow for the race condition you mentioned.